ignite/pkg/cosmosclient/cosmosclient_test.go

```diff
@@ -0,0 +1,378 @@
+package cosmosclient
+
+import (
+	"context"
+	"io"
+	"os"
+	"path"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/tendermint/p2p"
+	ctypes "github.com/tendermint/tendermint/rpc/core/types"
+
+	"github.com/ignite/cli/ignite/pkg/cosmosaccount"
+	"github.com/ignite/cli/ignite/pkg/cosmosclient/mocks"
+)
+
+//go:generate mockery --srcpkg github.com/tendermint/tendermint/rpc/client/ --name Client --structname RPCClient --filename rpclient.go --with-expecter
+
+type suite struct {
+	rpcClient *mocks.RPCClient
+}
+
+func newSuite(t *testing.T, setup func(suite)) suite {
+	s := suite{
+		rpcClient: mocks.NewRPCClient(t),
+	}
+	if setup != nil {
+		setup(s)
+	}
+	return s
+}
+
+func TestNew(t *testing.T) {
+	var (
+		ctx     = context.Background()
+		home, _ = os.UserHomeDir()
+	)
+	tests := []struct {
+		name           string
+		opts           []Option
+		expectedClient Client
+		expectedError  string
+		setup          func(suite)
+	}{
+		{
+			name: "ok: default values",
+			setup: func(s suite) {
+				s.rpcClient.EXPECT().Status(ctx).Return(&ctypes.ResultStatus{
+					NodeInfo: p2p.DefaultNodeInfo{Network: "mychain"},
+				}, nil)
+			},
```

Comment:
Once we enable black box testing that suggested in the previous comment, we should start initializing the client by using its new function because we no longer will be able to access private fields.

Reply:
I'm testing the `New` function actually, so I can't use it to build the expected result, or else I'm faking the test ;)

This is also why I couldn't use the `cosmosclient_test` package for the test. That said, if you think black box testing is really important, I can separate the tests in 2 files :
- `cosmosclient_test.go` contains only `TestNew` which requires to have access to the private fields using the `cosmosclient` package
- `cosmosclient_private_test.go` contains the remaining tests and uses the `cosmosclient_test` package.

Reply:
I think we should always avoid creating tests that depends on private declarations (implementation details).

Reply:
But the alternatives are not very attractives:
- don't test :fearful: 
- add getters and setters for each field :coffee: 

IMO, having something properly tested is much better than keeping a pattern like black box testing at all cost.

Sometimes it's just not possible to assert properly w/o accessing the private fields, or sometimes you have this very complicated private function and having a specific test on it is super helpful to ensure it works as expected.

Note that even in the stdlib, not all tests use black box testing.

Reply:
Hey @ilgooz I worked on a black box testing only version [cfaae9e](https://github.com/ignite/cli/pull/2789/commits/cfaae9e1baa50136277ef5ef9792e209cf6ec25e)

The caveat is some fields are no longer asserted, like `nodeAddress`. If you prefer like that, I can go with that too :)